File: src/main/java/com/matamercer/microblog/models/entities/BaseModel.java
package com.matamercer.microblog.models.entities;

import javax.persistence.*;
import java.io.Serializable;
import java.time.ZonedDateTime;
import lombok.Getter;
import lombok.Setter;
import org.apache.commons.lang3.builder.HashCodeBuilder;

@Getter
@Setter
@MappedSuperclass
public abstract class BaseModel implements Comparable<BaseModel>, Serializable {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Long id;

    @Column(nullable = false)
    private ZonedDateTime createdAt;

    @Column(nullable = false)
    private ZonedDateTime updatedAt;

    @PrePersist
    public void prePersist() {
        createdAt = updatedAt = ZonedDateTime.now();
    }

    @PreUpdate
    public void preUpdate() {
        updatedAt = ZonedDateTime.now();
    }

    @Override
    public int compareTo(BaseModel o) {
        return this.getId().compareTo(o.getId());
    }

    public boolean equals(Object other) {
        if (other == null || other.getClass() != this.getClass()) {
            return false;
        }

        return this.getId().equals(((BaseModel) other).getId());
    }


    public int hashCode() {
        return new HashCodeBuilder().append(getId()).toHashCode();
    }
}

File: src/main/java/com/matamercer/microblog/models/entities/BaseModel.kt
package com.matamercer.microblog.models.entities

import org.apache.commons.lang3.builder.HashCodeBuilder
import java.io.Serializable
import java.time.ZonedDateTime
import javax.persistence.*

@MappedSuperclass
abstract class BaseModel : Comparable<BaseModel>, Serializable {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    var id: Long? = null

    @Column(nullable = false)
    var createdAt: ZonedDateTime? = null

    @Column(nullable = false)
    var updatedAt: ZonedDateTime? = null
    @PrePersist
    fun prePersist() {
        updatedAt = ZonedDateTime.now()
        createdAt = updatedAt
    }

    @PreUpdate
    fun preUpdate() {
        updatedAt = ZonedDateTime.now()
    }

    override fun compareTo(o: BaseModel): Int {
        return o.id?.let { id?.compareTo(it) } ?: 0
    }

    override fun equals(other: Any?): Boolean {
        return if (other == null || other.javaClass != this.javaClass) {
            false
        } else this.id == (other as BaseModel).id
    }

    override fun hashCode(): Int {
        return HashCodeBuilder().append(id).toHashCode()
    }
}